File: ios/SampleApp/Screens/Login/LoginViewModel.swift
import SwiftUI

enum LoginState: Equatable {
    case idle
    case loading
    case success
    case error
}

@MainActor
final class LoginViewModel: ObservableObject {
    @Published private(set) var state: LoginState = .idle

    func login(email: String, password: String) {
        guard state != .loading else { return }

        state = .loading
        Task {
            let succeeded = await MockRepository.shared.login(email: email, password: password)
            state = succeeded ? .success : .error
        }
    }
}

File: android/app/src/main/java/com/example/sampleapp/ui/login/LoginViewModel.kt
package com.example.sampleapp.ui.login

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.sampleapp.data.MockRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

sealed interface LoginState {
    data object Idle : LoginState
    data object Loading : LoginState
    data object Success : LoginState
    data object Error : LoginState
}

class LoginViewModel : ViewModel() {
    private val _state = MutableStateFlow<LoginState>(LoginState.Idle)
    val state: StateFlow<LoginState> = _state.asStateFlow()

    fun login(email: String, password: String) {
        if (_state.value == LoginState.Loading) return

        viewModelScope.launch {
            _state.value = LoginState.Loading
            _state.value = if (MockRepository.login(email, password)) {
                LoginState.Success
            } else {
                LoginState.Error
            }
        }
    }
}